internal/core/metadata/rest_device.go

```diff
@@ -64,71 +66,47 @@ func restAddNewDevice(w http.ResponseWriter, r *http.Request) {
 	err := json.NewDecoder(r.Body).Decode(&d)
 	if err != nil {
 		LoggingClient.Error(err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
+		status := http.StatusBadRequest
+		if _, ok := err.(models.ErrContractInvalid); !ok {
+			status = http.StatusInternalServerError
+		}
+		http.Error(w, err.Error(), status)
 		return
 	}
 
-	// Protocol check
-	if len(d.Protocols) == 0 {
-		err := errors.New("no supporting protocol specified for device")
-		LoggingClient.Error(err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
+	ctx := r.Context()
+	// The following requester instance is necessary because we will be making an HTTP call to the device service
+	// associated with the new device in the Notifier below. There is no device service client. Additionally, the
+	// requester interface should be mocked for unit testability and so is injected into the Notifier.
+	requester, err := device.NewRequester(device.Http, LoggingClient, ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	// Service Check
-	// Try by name
-	service, err := dbClient.GetDeviceServiceByName(d.Service.Name)
-	if err != nil {
-		// Try by ID
-		service, err = dbClient.GetDeviceServiceById(d.Service.Id)
-		if err != nil {
-			LoggingClient.Error(err.Error())
-			http.Error(w, err.Error()+": A device must be associated with a device service", http.StatusBadRequest)
-			return
-		}
-	}
-	d.Service = service
+	ch := make(chan device.DeviceEvent)
+	defer close(ch)
 
-	// Profile Check
-	// Try by name
-	d.Profile, err = dbClient.GetDeviceProfileByName(d.Profile.Name)
-	if err != nil {
-		// Try by ID
-		d.Profile, err = dbClient.GetDeviceProfileById(d.Profile.Id)
-		if err != nil {
-			LoggingClient.Error(err.Error())
-			http.Error(w, err.Error()+": A device must be associated with a device profile", http.StatusBadRequest)
-			return
-		}
-	}
-
-	// Check operating/admin state
-	if d.OperatingState == models.OperatingState("") || d.AdminState == models.AdminState("") {
-		err = errors.New("Device can't have null operating state or admin state")
-		LoggingClient.Error(err.Error())
-		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
-	}
+	notifier := device.NewNotifier(ch, nc, Configuration.Notifications, dbClient, requester, LoggingClient, ctx)
+	go notifier.Execute()
```

Comment:
Might want to add comment that this will block waiting on the event channel.

Reply:
Unsure what you mean. It won't block execution because it's in a goroutine, but it will block within the goroutine itself -- which doesn't affect request handling.

Reply:
Yea, I meant that Execute will block waiting on the event channel. Just clarifies why it is being called at this point in the code.